File: rudder-client-ios/RudderClient/Event/RudderMessage.swift
//
//  RudderMessage.swift
//  RudderSample
//
//  Created by Arnab Pal on 10/07/19.
//  Copyright © 2019 Arnab Pal. All rights reserved.
//

import Foundation

class RudderMessage : Encodable {
    let messageId: String = UUID().uuidString.lowercased()
    var channel: String = ""
    let context : RudderContext = RudderContext()
    var type: String = ""
    let action: String = ""
    let timestamp: String = Date().toTimeStampString()
    var anonymousId: String = ""
    var userId: String = ""
    var event: String = ""
    var properties: RudderProperty = RudderProperty()
    var integrations: [String] = ["rudderlabs"]
    
    func initiate(template: RudderEventTemplate) {
        self.anonymousId = template.deviceId
        self.context.initiate(template: template)
    }
    
    func addIntegration(integration: String) {
        self.integrations.append(integration)
    }
    
    enum CodingKeys: String, CodingKey {
        case messageId = "rl_message_id"
        case channel = "rl_channel"
        case context = "rl_context"
        case type = "rl_type"
        case action = "rl_action"
        case timestamp = "rl_timestamp"
        case anonymousId = "rl_anonymous_id"
        case userId = "rl_user_id"
        case event = "rl_event"
        case properties = "rl_properties"
        case integrations = "rl_integrations"
    }
    
    func encode(to encoder: Encoder) throws {
        var container = encoder.container(keyedBy: CodingKeys.self)
        try container.encode(messageId, forKey: .messageId)
        try container.encode(channel, forKey: .channel)
        try container.encode(context, forKey: .context)
        try container.encode(type, forKey: .type)
        try container.encode(action, forKey: .action)
        try container.encode(timestamp, forKey: .timestamp)
        try container.encode(anonymousId, forKey: .anonymousId)
        try container.encode(userId, forKey: .userId)
        try container.encode(event, forKey: .event)
        try container.encode(properties.property, forKey: .properties)
        try container.encode(integrations, forKey: .integrations)
    }
}

File: rudder-android-client/rudder-sdk-core/src/main/java/com/rudderlabs/android/sdk/core/event/RudderMessage.java
package com.rudderlabs.android.sdk.core.event;

import com.google.gson.annotations.SerializedName;
import com.rudderlabs.android.sdk.core.util.Utils;

import java.util.ArrayList;
import java.util.Locale;
import java.util.Map;
import java.util.UUID;

class RudderMessage {
    @SerializedName("rl_channel")
    private String channel = "android-sdk";
    @SerializedName("rl_context")
    private RudderContext context;
    @SerializedName("rl_type")
    private String type;
    @SerializedName("rl_action")
    private String action;
    @SerializedName("rl_message_id")
    private String messageId = String.format(Locale.US, "%d-%s", System.currentTimeMillis(), UUID.randomUUID().toString());
    @SerializedName("rl_timestamp")
    private String timestamp = Utils.getTimeStamp();
    @SerializedName("rl_anonymous_id")
    private String anonymousId;
    @SerializedName("rl_user_id")
    private String userId;
    @SerializedName("rl_event")
    private String event;
    @SerializedName("rl_properties")
    private Map<String, Object> properties;
    @SerializedName("rl_user_properties")
    private Map<String, Object> userProperties;
    @SerializedName("rl_integrations")
    private ArrayList<String> integrations = new ArrayList<>();

    RudderMessage() {
        context = RudderElementCache.getCachedContext();
    }

    void setProperty(RudderProperty property) {
        if (property != null) this.properties = property.getMap();
    }

    void setUserProperties(RudderUserProperty userProperty) {
        this.userProperties = userProperty.getMap();
    }

    void setType(String type) {
        this.type = type;
    }

    void setIntegrations(ArrayList<String> integrations) {
        this.integrations.addAll(integrations);
    }

    public void setUserId(String userId) {
        this.userId = userId;
    }

    public void setEventName(String eventName) {
        this.event = eventName;
    }

    void updateTraits(RudderTraits traits) {
        this.context.updateTraits(traits);
    }
}
